### PR TITLE
[Fail Gracefully] Inform user when service is not configured so that the app does not fail/crash

### PR DESCRIPTION
**Pull request by ziccardi**
# Motivation
ISSUE https://issues.jboss.org/browse/AEROGEAR-3064

We want the user to have an error message instead of application crash if a service is not configured.

Screenshot can be found in the slideset, slides 12 (android) and 13 (iOS)

## Description

- Added ConstrainedNavigationItem so that a generic list of constraints can be added to a view. For each of those constraints a custom action can be executed.
- Added IConstraint interface, so that custom constraints can be added
- Added HasConfigurationConstraint to verify that the service configuration is present
- Added AuthNavigationItem to configure custom constraints (HasConfigurationConstraint) for the Auth Page
- Added an enumeration for all the items, so that we can jump from one page to the other by code

### CODE REVIEW

File: Example/Models/Constraints/HasConfigurationContraint.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        private async void DisplayAlert()
+        {
+            RootPage Root = (RootPage)(Xamarin.Forms.Application.Current.MainPage);
+            if (await Root.DisplayAlert(
+                StringResources.dialog_notconfigured_title,
+                String.Format(StringResources.dialog_notconfigured_message, serviceName),
+                StringResources.button_showdocs_caption,
+                StringResources.button_close_caption))
```
[the remark this answers is not in the document]
Comment (ziccardi) [reply Example/Models/Constraints/HasConfigurationContraint.cs]: Yes. If close is chosen, nothing happens

**Comment by twnolan** [comment]
@ziccardi Verified as working

File: Example.Android/MainActivity.cs [hunk 2]
```diff
@@ -36,9 +50,15 @@ protected override void OnCreate(Bundle savedInstanceState)
             var app = new App();
             MobileCoreAndroid.Init(app.GetType().Assembly,ApplicationContext);
             SecurityService.InitializeService();
-            IAuthService authService = AuthService.InitializeService();
-            var authConfig = AuthenticationConfig.Builder.RedirectUri("org.aerogear.mobile.example:/callback").Build();
-            authService.Configure(authConfig);
+
+            if (SilentlyRun(() => AuthService.InitializeService()))
```
Comment (wei-lee) [inline Example.Android/MainActivity.cs]: not sure I like this idea. Should we consider using the core module to get the configuration, or just use a try/catch here?

Comment (secondsun) [reply Example.Android/MainActivity.cs]: When you say the configuration are you talking about the mobile app config (mostly redirect ui) or are you talking about the service conifg (the keycloak urls etc)?

Comment (ziccardi) [reply Example.Android/MainActivity.cs]: @wei-lee @secondsun The issue here is that the InitializeService will throw an exception if the serviceconfiguration is not present.

If we manually check for the configuration to be present, then we must put a `custom` if for any service we want to check. Moreover, we need to put the knowledge of what configuration to be checked inside the app (currently that knowledge is mainly inside the service itself).

If we put a `try/catch` then we will need to put every service we add inside a try catch making the code cluttered: 
```csharp
try
{
   AuthService.InitializeService();
} 
catch 
{ // do nothing
}
try
{
   OtherService.InitializeService();
}
catch
{  //do nothing
}
```

I believe the `SilentlyRun` approach is the cleanest solution (we can log the exception).
Maybe we can rename it to `FailSafeRun`.

Comment (secondsun) [reply Example.Android/MainActivity.cs]: Oh wait, you mean for if the service config is missing.  Derp

Comment (wei-lee) [reply Example.Android/MainActivity.cs]: @ziccardi ok, right, I think I understand now. Thanks for the explanation.

**Pull request merged**